src/Engine/Objects/Actor.cpp

```diff
@@ -4129,117 +4129,66 @@ void toggleActorGroupFlag(unsigned int uGroupID, ActorAttribute uFlag,
+        pActors[i].ResetFullAiState();
+        if (!pActors[i].CanAct()) {
+            pActors[i].ResetActive();
             continue;
         }
 
-        v22 = abs(pParty->vPosition.z - actor->vPosition.z);
-        v21 = abs(pParty->vPosition.y - actor->vPosition.y);
-        v1 = abs(pParty->vPosition.x - actor->vPosition.x);
-        v7 = int_get_vector_length(v22, v21, v1);
-        distance = v7 - actor->uActorRadius;
-        if (distance < 0) distance = 0;
+        int delta_x = abs(pParty->vPosition.x - pActors[i].vPosition.x);
+        int delta_y = abs(pParty->vPosition.y - pActors[i].vPosition.y);
+        int delta_z = abs(pParty->vPosition.z - pActors[i].vPosition.z);
+
+        int distance = int_get_vector_length(delta_x, delta_y, delta_z) - pActors[i].uActorRadius;
+        if (distance < 0)
+            distance = 0;
 
         if (distance < 5632) {
-            actor->ResetHostile();
-            if (actor->ActorEnemy() || actor->GetActorsRelation(0)) {
-                // v11 = (pParty->uFlags & 0x10) == 0;
-                actor->uAttributes |= ACTOR_HOSTILE;
-                if (distance < 5120) pParty->SetYellowAlert();
-                if (distance < 307) pParty->SetRedAlert();
+            pActors[i].ResetHostile();
+            if (pActors[i].ActorEnemy() || pActors[i].GetActorsRelation(0)) {
+                pActors[i].uAttributes |= ACTOR_HOSTILE;
+                if (distance < 5120)
+                    pParty->SetYellowAlert();
+                if (distance < 307)
+                    pParty->SetRedAlert();
             }
-            actor->uAttributes |= ACTOR_ACTIVE;
-            ai_near_actors_distances[ai_arrays_size] = distance;
-            ai_near_actors_ids[ai_arrays_size++] = i;
+            pActors[i].uAttributes |= ACTOR_ACTIVE;
+            activeActorsDistances.push_back({i, distance});
         } else {
-            actor->ResetActive();
+            pActors[i].ResetActive();
         }
     }
 
-    /*
-    result = v27;
-    if ( v27 > 0 )
-    {
-      v14 = 0;
-      v15 = 1;
-      v26 = 1;
-      do
-      {
-        while ( 1 )
-        {
-          v24 = v15;
-          if ( v15 >= result )
-            break;
-          v16 = ai_near_actors_distances[v14];
-          if ( v16 > ai_near_actors_distances[v15] )
-          {
-            v17 = &ai_near_actors_ids[v15];
-            v18 = ai_near_actors_ids[v14];
-            ai_near_actors_ids[v14] = *v17;
-            *v17 = v18;
-            v15 = v24;
-            ai_near_actors_distances[v14] = ai_near_actors_distances[v24];
-            ai_near_actors_distances[v24] = v16;
-          }
-          result = v27;
-          ++v15;
-        }
-        ++v14;
-        v15 = v26 + 1;
-        v26 = v15;
-      }
-      while ( v15 - 1 < result );
-    }*/
-
+    // sort active actors by distance
+    std::sort(activeActorsDistances.begin(), activeActorsDistances.end(), [] (std::pair<int, int> a, std::pair<int, int> b) {
```

Comment:
`std::nth_element` should do the job here

Reply:
You lost me here :(

Reply:
I'm saying that you're only really using the first 30 elements there, thus you don't need the array to be sorted, you just need to have it partitioned.